lua/ttt2/libraries/marker_vision.lua

```diff
@@ -21,160 +22,203 @@ markerVision = {}
--- @param[opt] any passThroughData any data that should be added to this radar vision, it is also synced to the client
--- @note Call on server to add entity on server and all defined clients.
+-- Creates a new marker vision object for the entity.
+-- @note This does not sync to the client, @{MARKER_VISION_OBJECT:SyncToClients}
+-- has to be called on it first
+-- @param Entity ent The entity which should receive the marker vision element
+-- @param string identifier The unique identifier of this marker vision element
+-- @return MARKER_VISION_OBJECT The marker vision object that was created
 -- @realm shared
-function markerVision.RegisterEntity(ent, owner, visibleFor, color, receiverList, passThroughData)
-	if SERVER then
-		local plysTeam = GetTeamFilter(owner:GetTeam(), true, true)
-
-		if #plysTeam == 0 then -- happens for TEAM_NONE for example
-			visibleFor = VISIBLE_FOR_PLAYER
-		end
+function markerVision.Add(ent, identifier)
+	local mvObject = table.Copy(MARKER_VISION_OBJECT)
+	mvObject:SetEnt(ent)
+	mvObject:SetIdentifier(identifier)
 
-		local streamTable = {
-			ent = ent,
-			owner = owner,
-			visibleFor = visibleFor,
-			color = color,
-			data = passThroughData
-		}
+	markerVision.registry[#markerVision.registry + 1] = mvObject
 
-		local receiver
-
-		if visibleFor == VISIBLE_FOR_PLAYER then
-			receiver = owner
-		elseif visibleFor == VISIBLE_FOR_TEAM then
-			receiver = plysTeam
-		end
-
-		-- note: when VISIBLE_FOR_ALL is used, the receiver will be nil and therefore SendStream uses net.Broadcast
-		net.SendStream("ttt2_register_entity_added", streamTable, receiverList or receiver)
-	end
-
-	if CLIENT then
-		if visibleFor == VISIBLE_FOR_ALL then
-			color = color or TEAMS[TEAM_INNOCENT].color
-		elseif visibleFor == VISIBLE_FOR_PLAYER then
-			color = color or TEAMS[TEAM_NONE].color
-		else
-			color = color or TEAMS[LocalPlayer():GetTeam()].color
-		end
-
-		marks.Add({ent}, color)
-	end
-
-	markerVision.registry[ent] = {
-		owner = owner,
-		visibleFor = visibleFor,
-		color = color,
-		data = passThroughData
-	}
+	return mvObject
 end
 
 ---
--- Returns the visibleFor flag for a provided entity, nil if not set
--- @param Entity ent The tracked entity
--- @return number|nil The visibility flag
+-- Returns the marker vision element if it exists.
+-- @param Entity ent The entity which should receive the marker vision element
+-- @param string identifier The unique identifier of this marker vision element
+-- @return MARKER_VISION_OBJECT The marker vision object
 -- @realm shared
-function markerVision.GetVisibleFor(ent)
-	if not IsValid(ent) or not markerVision.registry[ent] then return end
+function markerVision.Get(ent, identifier)
+	for i = 1, #markerVision.registry do
+		local mvObject = markerVision.registry[i]
 
-	return markerVision.registry[ent].visibleFor
+		if mvObject:IsSearchedObject(ent, identifier) then
+			return mvObject, i
+		end
+	end
```

Comment:
Personally, i would have used a 2-dimensional lookup table from ent -> identifier -> mvObject instead. 

But this is rather an optional suggestion.

Reply:
My main concern was performance when rendering. Therefore I chose this structure

Reply:
Table lookups are faster though

Reply:
I could cache the table index on the object. That way we would have the best performance imho

Reply:
This is executed on every frame. It is always advised against using pairs loops in rendering. Which I would have to do if I'm using your suggestion

If I'd use two sequential tables together, I don't see any benefit except having a more complex structure for no real reason

Reply:
I think you dont get what im saying here. I suggest using a table indexed by entity and then id. You are currently looping over it

Reply:
Yes, I did that previously but this would mean two pairs loops combined which I really don't want to have inside a rendering hook

Reply:
No it does not its a Lookup Table that means -> registry[entitiy][identifier] and you got it

Reply:
I know. But I have to iterate over it each frame. Or should I have two tables? One for rendering, one for lookup?

Reply:
First of: 2 loops are not really an issue for such code. This would really be a micro optimization i believe, but as i said this is an optional suggestion so will close this.